src/main/java/org/icgc/argo/program_service/services/ProgramService.java

```diff
@@ -138,14 +128,33 @@ public Boolean acceptInvite(UUID invitationId) {
     }
   }
 
-  public void removeProgram(UUID programId) {
-    val program = programRepository.findById(programId);
-    if (program.isPresent()) {
-      removeProgram(program.get());
-    } else {
-      //TODO: add proper error handling
-      log.error("Could not find program {}", programId);
+  public ErrorOr<ProgramEntity> createProgram(@NonNull Program program) {
+    val programEntity = programConverter.programToProgramEntity(program);
+    val now = LocalDateTime.now(ZoneId.of("UTC"));
+    programEntity.setCreatedAt(now);
+    programEntity.setUpdatedAt(now);
+
+    ProgramEntity saved;
+    try {
+      saved = programRepository.save(programEntity);
+    } catch(IOError error) {
+      return new ErrorOr(error);
+    } catch(NoSuchElementException e) {
+      return new ErrorOr(new RuntimeException("Could not save program"));
+    } catch(NullPointerException e) {
+      return new ErrorOr(new Error("Could not save program"));
+    }
+    if (saved == null) {
+      return new ErrorOr(new Error("Could not save program -- no idea why..."));
     }
+
+    try {
+      egoService.setUpProgram(saved);
+    } catch(HttpClientErrorException error) {
+      return new ErrorOr(error);
+    }
+
+    return new ErrorOr(saved);
```

Comment:
The issue with returning an ErrorOr is that it swallows the exception. This method should be transactional, so that if setupProgram fails, the save that previously happened in programRepository should be rolled back. By default, rolling back is triggered by a RuntimeExceptions, so this try catch would hide the transaction. As a guideline, in methods where transactions are being called, there should only be try catches for recoverable exceptions.